### ISSUE TITLE
not sure why object r not found'

### ISSUE THREAD

**Issue by AnnaGoodman1** [ropensci/stplanr#310]
# SET UP
rm(list = ls())
library(foreign)
library(tidyverse)
library(rgdal)
library(stplanr)

memory.limit(size=1000000)

# Define cents and lines
cents_all <- readOGR(file.path("0_DataInput/lsoa_cents/lsoa_cents_mod.geojson"))
# https://github.com/npct/pct-inputs/blob/master/02_intermediate/01_geographies/lsoa_cents_mod.geojson")
lines_toroute_data <- readRDS(file.path("1_DataCreated/lines_toroute_data.Rds"))
# https://github.com/metahit/mh-route-commutes/blob/master/1_DataCreated/lines_toroute_data.Rds

# MAKE A SPATIAL OBJECT OF LINES
lines_toroute_lines <- od2line(flow = lines_toroute_data, zones = cents_all, destinations = cents_all) 
rownames(lines_toroute_data) <- sapply(1:length(lines_toroute_lines), function(j) lines_toroute_lines@lines[[j]]@ID) # FORCE DATA ROW NAMES TO BE SAME AS ID IN LINES (in case don't start from '1')
lines_toroute <- SpatialLinesDataFrame(sl = lines_toroute_lines, data = lines_toroute_data)
proj_4326 <- CRS("+proj=longlat +init=epsg:4326")   # global projection - lat/long.
lines_toroute <- spTransform(lines_toroute, proj_4326)

# TRY TO ROUTE IN CYCLE STREETS
route_type <- "fastest" # run for fastest then quietest
routes <- line2route(lines_toroute, route_fun = route_cyclestreet, plan = route_type, n_processes = 10, base_url = "http://pct.cyclestreets.net/api/")

**Comment by AnnaGoodman1** [comment]
in above code, I get an error message:

![image](https://user-images.githubusercontent.com/11818225/57627495-0029f180-7590-11e9-8721-c8c220120101.png)

**Comment by AnnaGoodman1** [comment]
(NB the 2 input datasets I show where to download them from, not sure how to get the R code to take them directly from git, sorry)

**Comment by Robinlovelace** [comment]
Thanks for the issue. Could you provide a reproducible issue report? You could start a script with `download.file()`. Also, I find the `reprex` package really useful: you can just enter something like:

```r
reprex::reprex({
x = "this is some test code"
print(x)
mean(x) # this is a bug
})
```

Into R and it will copy the nicely formatted results that can be pasted like this, to show the error message in a reproducible way that's easy on the :eye: :+1: : 

``` r
x = "this is some test code"
print(x)
#> [1] "this is some test code"
mean(x) # this is a bug
#> Warning in mean.default(x): argument is not numeric or logical: returning
#> NA
#> [1] NA
```

<sup>Created on 2019-05-13 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

Thanks for raising the issue in any case, could lead to improvements in `od2line()` I suspect, if only I can reproduce it quickly.

**Comment by AnnaGoodman1** [comment]
sorry, tried but e.g. still not managing to download the files straight from git, and struggling with other bits of doing code that way

**Comment by AnnaGoodman1** [comment]
it is easier on eye here... https://github.com/metahit/mh-route-commutes/blob/master/2.1_MergeBikeRoutes_minimal.R

**Comment by Robinlovelace** [comment]
I've tried to reproduce it on a sample of 10 lines in the Isle of Wight and it seems to work fine. Can you reproduce this? If so, I suspect it may be an issue with the input data. Also note the package versions in the details, suggest trying on updated packages if not up-to-day (print `Sys.Info()` should show what versions you have): 

``` r
# Aim: test line2route
    
    # ensure packages are installed and up-to-date
    install.packages("remotes")
#> Installing package into '/usr/local/lib/R/site-library'
#> (as 'lib' is unspecified)
    pkgs = c(
        "pct",
        "sf",
        "stplanr",
        "tidyverse",
        "tmap"
    )
    remotes::install_cran(pkgs)
#> Skipping install of 'pct' from a cran remote, the SHA1 (0.1.3) has not changed since last install.
#>   Use `force = TRUE` to force installation
#> Skipping install of 'sf' from a cran remote, the SHA1 (0.7-4) has not changed since last install.
#>   Use `force = TRUE` to force installation
#> Skipping install of 'stplanr' from a cran remote, the SHA1 (0.2.9) has not changed since last install.
#>   Use `force = TRUE` to force installation
#> Skipping install of 'tidyverse' from a cran remote, the SHA1 (1.2.1) has not changed since last install.
#>   Use `force = TRUE` to force installation
#> Skipping install of 'tmap' from a cran remote, the SHA1 (2.2) has not changed since last install.
#>   Use `force = TRUE` to force installation
    
    l = pct::get_pct_lines("isle-of-wight")
#> Loading required package: sp
    l_top_10 = dplyr::top_n(l, 10, wt = all)
    r = stplanr::line2route(l_top_10)
#> 10 % out of 10 distances calculated
#> 20 % out of 10 distances calculated
#> 30 % out of 10 distances calculated
#> 40 % out of 10 distances calculated
#> 50 % out of 10 distances calculated
#> 60 % out of 10 distances calculated
#> 70 % out of 10 distances calculated
#> 80 % out of 10 distances calculated
#> 90 % out of 10 distances calculated
#> 100 % out of 10 distances calculated
    plot(r)
#> Warning: plotting the first 10 out of 17 attributes; use max.plot = 17 to
#> plot all
```

![](https://i.imgur.com/8oq8MZz.png)

``` r
    
    l_sp = as(l_top_10, "Spatial")
    r = stplanr::line2route(l_sp)
#> 10 % out of 10 distances calculated
#> 20 % out of 10 distances calculated
#> 30 % out of 10 distances calculated
#> 40 % out of 10 distances calculated
#> 50 % out of 10 distances calculated
#> 60 % out of 10 distances calculated
#> 70 % out of 10 distances calculated
#> 80 % out of 10 distances calculated
#> 90 % out of 10 distances calculated
#> 100 % out of 10 distances calculated
    plot(r)
```

![](https://i.imgur.com/lXXgOG4.png)

<sup>Created on 2019-05-13 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>
<details><summary>Session info</summary>

``` r
devtools::session_info()
#> ─ Session info ──────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 3.5.2 (2018-12-20)
#>  os       Debian GNU/Linux 9 (stretch)
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       Etc/UTC                     
#>  date     2019-05-13                  
#> 
#> ─ Packages ──────────────────────────────────────────────────────────────
#>  package     * version date       lib source                          
#>  assertthat    0.2.1   2019-03-21 [1] CRAN (R 3.5.2)                  
#>  backports     1.1.4   2019-04-10 [1] CRAN (R 3.5.2)                  
#>  callr         3.2.0   2019-03-15 [1] CRAN (R 3.5.2)                  
#>  class         7.3-15  2019-01-01 [1] CRAN (R 3.5.2)                  
#>  classInt      0.3-3   2019-04-26 [1] CRAN (R 3.5.2)                  
#>  cli           1.1.0   2019-03-19 [1] CRAN (R 3.5.2)                  
#>  codetools     0.2-16  2018-12-24 [1] CRAN (R 3.5.2)                  
#>  crayon        1.3.4   2017-09-16 [1] CRAN (R 3.5.2)                  
#>  curl          3.3     2019-01-10 [1] CRAN (R 3.5.2)                  
#>  DBI           1.0.0   2018-05-02 [1] CRAN (R 3.5.2)                  
#>  desc          1.2.0   2018-05-01 [1] CRAN (R 3.5.2)                  
#>  devtools      2.0.2   2019-04-08 [1] CRAN (R 3.5.2)                  
#>  digest        0.6.18  2018-10-10 [1] CRAN (R 3.5.2)                  
#>  dplyr         0.8.0.1 2019-02-15 [1] CRAN (R 3.5.2)                  
#>  e1071         1.7-1   2019-03-19 [1] CRAN (R 3.5.2)                  
#>  evaluate      0.13    2019-02-12 [1] CRAN (R 3.5.2)                  
#>  foreign       0.8-71  2018-07-20 [2] CRAN (R 3.5.2)                  
#>  fs            1.3.1   2019-05-06 [1] CRAN (R 3.5.2)                  
#>  geosphere     1.5-7   2017-11-05 [1] CRAN (R 3.5.2)                  
#>  glue          1.3.1   2019-03-12 [1] CRAN (R 3.5.2)                  
#>  highr         0.8     2019-03-20 [1] CRAN (R 3.5.2)                  
#>  htmltools     0.3.6   2017-04-28 [1] CRAN (R 3.5.2)                  
#>  httr          1.4.0   2018-12-11 [1] CRAN (R 3.5.2)                  
#>  igraph        1.2.4.1 2019-04-22 [1] CRAN (R 3.5.2)                  
#>  jsonlite      1.6     2018-12-07 [1] CRAN (R 3.5.2)                  
#>  KernSmooth    2.23-15 2015-06-29 [2] CRAN (R 3.5.2)                  
#>  knitr         1.22    2019-03-08 [1] CRAN (R 3.5.2)                  
#>  lattice       0.20-38 2018-11-04 [2] CRAN (R 3.5.2)                  
#>  magrittr      1.5     2014-11-22 [1] CRAN (R 3.5.2)                  
#>  maptools      0.9-5   2019-02-18 [1] CRAN (R 3.5.2)                  
#>  memoise       1.1.0   2017-04-21 [1] CRAN (R 3.5.2)                  
#>  mime          0.6     2018-10-05 [1] CRAN (R 3.5.2)                  
#>  openxlsx      4.1.0   2018-05-26 [1] CRAN (R 3.5.2)                  
#>  pct           0.1.3   2019-05-13 [1] local                           
#>  pillar        1.4.0   2019-05-11 [1] CRAN (R 3.5.2)                  
#>  pkgbuild      1.0.3   2019-03-20 [1] CRAN (R 3.5.2)                  
#>  pkgconfig     2.0.2   2018-08-16 [1] CRAN (R 3.5.2)                  
#>  pkgload       1.0.2   2018-10-29 [1] CRAN (R 3.5.2)                  
#>  prettyunits   1.0.2   2015-07-13 [1] CRAN (R 3.5.2)                  
#>  processx      3.3.0   2019-03-10 [1] CRAN (R 3.5.2)                  
#>  ps            1.3.0   2018-12-21 [1] CRAN (R 3.5.2)                  
#>  purrr         0.3.2   2019-03-15 [1] CRAN (R 3.5.2)                  
#>  R.methodsS3   1.7.1   2016-02-16 [1] CRAN (R 3.5.2)                  
#>  R.oo          1.22.0  2018-04-22 [1] CRAN (R 3.5.2)                  
#>  R.utils       2.8.0   2019-02-14 [1] CRAN (R 3.5.2)                  
#>  R6            2.4.0   2019-02-14 [1] CRAN (R 3.5.2)                  
#>  raster        2.9-1   2019-02-26 [1] Github (rspatial/raster@9d05c09)
#>  Rcpp          1.0.1   2019-03-17 [1] CRAN (R 3.5.2)                  
#>  remotes       2.0.4   2019-04-10 [1] CRAN (R 3.5.2)                  
#>  rgdal         1.4-3   2019-03-14 [1] CRAN (R 3.5.2)                  
#>  rgeos         0.4-3   2019-04-24 [1] CRAN (R 3.5.2)                  
#>  rlang         0.3.4   2019-04-07 [1] CRAN (R 3.5.2)                  
#>  rmarkdown     1.12    2019-03-14 [1] CRAN (R 3.5.2)                  
#>  rprojroot     1.3-2   2018-01-03 [1] CRAN (R 3.5.2)                  
#>  sessioninfo   1.1.1   2018-11-05 [1] CRAN (R 3.5.2)                  
#>  sf            0.7-4   2019-04-25 [1] CRAN (R 3.5.2)                  
#>  sp          * 1.3-1   2018-06-05 [1] CRAN (R 3.5.2)                  
#>  stplanr       0.2.9   2019-05-09 [1] local                           
#>  stringi       1.4.3   2019-03-12 [1] CRAN (R 3.5.2)                  
#>  stringr       1.4.0   2019-02-10 [1] CRAN (R 3.5.2)                  
#>  testthat      2.1.1   2019-04-23 [1] CRAN (R 3.5.2)                  
#>  tibble        2.1.1   2019-03-16 [1] CRAN (R 3.5.2)                  
#>  tidyselect    0.2.5   2018-10-11 [1] CRAN (R 3.5.2)                  
#>  units         0.6-3   2019-05-03 [1] CRAN (R 3.5.2)                  
#>  usethis       1.5.0   2019-04-07 [1] CRAN (R 3.5.2)                  
#>  withr         2.1.2   2018-03-15 [1] CRAN (R 3.5.2)                  
#>  xfun          0.6     2019-04-02 [1] CRAN (R 3.5.2)                  
#>  xml2          1.2.0   2018-01-24 [1] CRAN (R 3.5.2)                  
#>  yaml          2.2.0   2018-07-25 [1] CRAN (R 3.5.2)                  
#>  zip           2.0.1   2019-03-11 [1] CRAN (R 3.5.2)                  
#> 
#> [1] /usr/local/lib/R/site-library
#> [2] /usr/local/lib/R/library
```

</details>


In summary I cannot repoduce on a minimal dataset so closing for now but happy to re-open if I can reproduce the error message.

**Comment by Robinlovelace** [comment]
Checking with data provided:

``` r
cents = sf::read_sf("E:/tmp/lsoa_cents_mod.geojson")
od_data = readRDS("E:/tmp/lines_toroute_data.Rds")
l_sf = stplanr::od2line(od_data, cents)
routes = stplanr::line2route(l_sf[1:20, ])
#> 10 % out of 20 distances calculated
#> 20 % out of 20 distances calculated
#> 30 % out of 20 distances calculated
#> 40 % out of 20 distances calculated
#> 50 % out of 20 distances calculated
#> 60 % out of 20 distances calculated
#> 70 % out of 20 distances calculated
#> 80 % out of 20 distances calculated
#> 90 % out of 20 distances calculated
#> 100 % out of 20 distances calculated
plot(routes)
#> Warning: plotting the first 9 out of 17 attributes; use max.plot = 17 to
#> plot all
```

![](https://i.imgur.com/bqlU4iM.png)

<sup>Created on 2019-05-14 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>
